// @clutch-sh/core/dist/Theme/theme.js
import React from "react";
import { ThemeProvider as EmotionThemeProvider } from "@emotion/react";
import { createMuiTheme, CssBaseline, MuiThemeProvider } from "@material-ui/core";
import { StylesProvider } from "@material-ui/styles";
import { clutchColors } from "./colors";
import palette from "./palette";
// Create a Material UI theme is propagated to all children.
const createTheme = (variant) => {
    return createMuiTheme({
        // inject in custom colors
        ...clutchColors(variant),
        palette: palette(variant),
        transitions: {
            // https://material-ui.com/getting-started/faq/#how-can-i-disable-transitions-globally
            create: () => "none",
        },
        props: {
            MuiButtonBase: {
                // https://material-ui.com/getting-started/faq/#how-can-i-disable-the-ripple-effect-globally
                disableRipple: true,
            },
        },
        overrides: {
            MuiAccordion: {
                root: {
                    "&$expanded": {
                        // remove the additional margin rule when expanded so the original margin is used.
                        margin: null,
                    },
                },
            },
        },
    });
};
const ThemeProvider = ({ children, variant = "light" }) => (React.createElement(MuiThemeProvider, { theme: createTheme(variant) },
    React.createElement(EmotionThemeProvider, { theme: createTheme(variant) },
        React.createElement(CssBaseline, null),
        React.createElement(StylesProvider, { injectFirst: true }, children))));
// Note that ThemeProvider can't be used until the Theme component can be replaced.
export default ThemeProvider;
