// @clutch-sh/core/dist/Feedback/alert.js
import * as React from "react";
import { Grid } from "@material-ui/core";
import MuiSuccessIcon from "@material-ui/icons/CheckCircle";
import MuiErrorIcon from "@material-ui/icons/Error";
import MuiInfoIcon from "@material-ui/icons/Info";
import MuiWarningIcon from "@material-ui/icons/Warning";
import { Alert as MuiAlert, AlertTitle as MuiAlertTitle } from "@material-ui/lab";
import styled from "../styled";
const backgroundColors = {
    error: "linear-gradient(to right, #DB3615 8px, #FDE9E7 0%)",
    info: "linear-gradient(to right, #3548D4 8px, #EBEDFB 0%)",
    success: "linear-gradient(to right, #1E942E 8px, #E6F7EB 0%)",
    warning: "linear-gradient(to right, #FFCC80 8px, #FFFDE6 0%)",
};
const StyledAlert = styled(MuiAlert)({
    borderRadius: "8px",
    padding: "16px",
    paddingLeft: "24px",
    paddingBottom: "20px",
    color: "rgba(13, 16, 48, 0.6)",
    fontSize: "14px",
    overflow: "auto",
    ".MuiAlert-icon": {
        marginRight: "16px",
        padding: "0",
    },
    ".MuiAlert-message": {
        maxWidth: "calc(100% - 40px)",
        width: "100%",
        padding: "0",
        ".MuiAlertTitle-root": {
            marginBottom: "0",
            color: "#0D1030",
        },
    },
}, props => ({
    background: backgroundColors[props.severity],
}));
const ErrorIcon = styled(MuiErrorIcon)({
    color: "#db3716",
});
const InfoIcon = styled(MuiInfoIcon)({
    color: "#3548d4",
});
const SuccessIcon = styled(MuiSuccessIcon)({
    color: "#1e942d",
});
const WarningIcon = styled(MuiWarningIcon)({
    color: "#ffcc80",
});
const AlertTitle = styled(MuiAlertTitle)({
    color: "#0D1030",
    fontWeight: 600,
    fontSize: "16px",
});
const iconMappings = {
    error: React.createElement(ErrorIcon, null),
    info: React.createElement(InfoIcon, null),
    success: React.createElement(SuccessIcon, null),
    warning: React.createElement(WarningIcon, null),
};
export const Alert = ({ severity = "info", title, children, ...props }) => (React.createElement(StyledAlert, Object.assign({ severity: severity, iconMapping: iconMappings }, props),
    title && React.createElement(AlertTitle, null, title),
    children));
export const AlertPanel = ({ direction = "column", children }) => (React.createElement(Grid, { container: true, direction: direction, justify: "center", alignContent: "space-between", wrap: "nowrap" }, children));
