// @clutch-sh/core/dist/Feedback/tooltip.js
import * as React from "react";
import styled from "@emotion/styled";
import { Tooltip as MuiTooltip } from "@material-ui/core";
const BaseTooltip = ({ className, ...props }) => (React.createElement(MuiTooltip, Object.assign({ classes: { tooltip: className } }, props)));
// TODO: sync with Design on margins for each possible placement
const StyledTooltip = styled(BaseTooltip)((props) => ({
    maxWidth: props.maxwidth,
    backgroundColor: "#0D1030",
    borderRadius: "6px",
    "&.MuiTooltip-tooltipPlacementLeft": {
        margin: "0 2px",
    },
    "&.MuiTooltip-tooltipPlacementRight": {
        margin: "0 2px",
    },
    "&.MuiTooltip-tooltipPlacementTop": {
        margin: "2px 0",
    },
    "&.MuiTooltip-tooltipPlacementBottom": {
        margin: "2px 0",
    },
}));
const Tooltip = ({ children, maxWidth = "300px", title, ...props }) => {
    return (React.createElement(StyledTooltip, Object.assign({ title: title, maxwidth: maxWidth }, props), children));
};
// sets the spacing between multiline content
const TooltipContainer = styled.div({
    margin: "4px 0px",
});
export { Tooltip, TooltipContainer };
