// @clutch-sh/core/dist/dialog.js
import * as React from "react";
import styled from "@emotion/styled";
import { Dialog as MuiDialog, DialogActions as MuiDialogActions, DialogContent as MuiDialogContent, DialogTitle as MuiDialogTitle, IconButton as MuiIconButton, Paper, } from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
const DialogPaper = styled(Paper)({
    border: "1px solid rgba(13, 16, 48, 0.1)",
    boxShadow: "0px 10px 24px rgba(35, 48, 143, 0.3)",
    boxSizing: "border-box",
    backgroundColor: "#FFFFFF",
    width: "max-content",
    maxWidth: "75vw",
});
const DialogTitle = styled(MuiDialogTitle)({
    display: "flex",
    justifyContent: "space-between",
    fontSize: "20px",
    padding: "12px 12px 0 32px",
    fontWeight: 500,
    color: "#0D1030",
});
const DialogTitleText = styled.div({
    padding: "14px 0 0 0",
});
const IconButton = styled(MuiIconButton)({
    height: "12px",
    width: "12px",
    color: "#0D1030",
});
const DialogContent = styled(MuiDialogContent)({
    padding: "16px 32px 32px 32px",
    fontSize: "16px",
    fontWeight: 400,
    color: "rgba(13, 16, 48, 0.6)",
    "> *": {
        margin: "16px 0 0 0",
    },
    overflowWrap: "break-word",
});
const DialogActions = styled(MuiDialogActions)({
    borderTop: "1px solid rgba(13, 16, 48, 0.12)",
    padding: "0 8px",
    "> *": {
        margin: "16px 8px 16px 8px",
    },
});
const Dialog = ({ title, children, open, onClose }) => (React.createElement(MuiDialog, { PaperComponent: DialogPaper, open: open, onClose: onClose },
    React.createElement(DialogTitle, { disableTypography: true },
        React.createElement(DialogTitleText, null, title),
        React.createElement(IconButton, { onClick: e => onClose(e, "closeButtonClick") },
            React.createElement(CloseIcon, null))),
    children));
export { Dialog, DialogActions, DialogContent };
