// @clutch-sh/core/dist/horizontal-rule.js
import * as React from "react";
import styled from "@emotion/styled";
const HorizontalRuleBase = ({ children, ...props }) => (React.createElement("div", Object.assign({}, props),
    React.createElement("div", { className: "line" },
        React.createElement("span", null)),
    React.Children.count(children) > 0 && React.createElement("div", { className: "content" }, children),
    React.createElement("div", { className: "line" },
        React.createElement("span", null))));
const StyledHorizontalRule = styled(HorizontalRuleBase)({
    alignItems: "center",
    display: "flex",
    flexDirection: "row",
    width: "100%",
    margin: "16px 0px",
    ".line": {
        flex: "1 1 auto",
    },
    ".line > span": {
        display: "block",
        borderTop: "1px solid rgba(13, 16, 48, 0.12)",
    },
    ".content": {
        padding: "0 16px",
        fontWeight: "bold",
        fontSize: "14px",
        color: "rgba(13, 16, 48, 0.38)",
        textTransform: "uppercase",
        display: "inline-flex",
        alignItems: "center",
    },
});
export const HorizontalRule = ({ children }) => (React.createElement(StyledHorizontalRule, null, children));
export default HorizontalRule;
