// @clutch-sh/core/dist/popper.js
import * as React from "react";
import styled from "@emotion/styled";
import { ClickAwayListener, Collapse, List, ListItem as MuiListItem, ListItemText as MuiListItemText, Paper as MuiPaper, Popper as MuiPopper, } from "@material-ui/core";
const StyledPopper = styled(MuiPopper)({
    zIndex: 1201,
    paddingTop: "16px",
});
const Paper = styled(MuiPaper)({
    minWidth: "fit-content",
    border: "1px solid #E7E7EA",
    boxShadow: "0px 10px 24px rgba(35, 48, 143, 0.3)",
    ".MuiListItem-root[id='popperItem']": {
        backgroundColor: "#FFFFFF",
        height: "48px",
        "&:hover": {
            backgroundColor: "#F5F6FD",
        },
        "&:active": {
            backgroundColor: "#D7DAF6",
        },
        "&.Mui-selected": {
            backgroundColor: "#FFFFFF",
            "&:hover": {
                backgroundColor: "#F5F6FD",
            },
            "&:active": {
                backgroundColor: "#D7DAF6",
            },
        },
        "&:hover, &:active, &.Mui-selected": {
            ".MuiTypography-root": {
                color: "#3548D4",
            },
        },
    },
});
const ListItem = styled(MuiListItem)({
    padding: "0",
});
const PopperItemIcon = styled.div({
    margin: "12px 5px 12px 12px",
    height: "24px",
    width: "24px",
});
const ListItemText = styled(MuiListItemText)({
    ".MuiTypography-root": {
        color: "rgba(13, 16, 48, 0.6)",
        fontWeight: 500,
        fontSize: "14px",
        lineHeight: "18px",
        padding: "15px 15px",
    },
});
const PopperItem = ({ children, componentProps, onClick, icon, ...props }) => (React.createElement(ListItem, Object.assign({ button: true, onClick: onClick, id: "popperItem", dense: true }, props, componentProps),
    icon && React.createElement(PopperItemIcon, null, icon),
    React.createElement(ListItemText, null, children)));
const Popper = ({ open, anchorRef, onClickAway, placement = "right-start", children, id, }) => (React.createElement(Collapse, { in: open, timeout: "auto", unmountOnExit: true },
    React.createElement(StyledPopper, { open: open, anchorEl: anchorRef === null || anchorRef === void 0 ? void 0 : anchorRef.current, transition: true, placement: placement },
        React.createElement(Paper, null,
            React.createElement(ClickAwayListener, { onClickAway: onClickAway },
                React.createElement(List, { component: "div", disablePadding: true, id: id }, children))))));
export { Popper, PopperItem };
