// @clutch-sh/core/dist/typography.js
import * as React from "react";
import styled from "./styled";
const REGULAR = 400;
const MEDIUM = 500;
const BOLD = 700;
const STYLE_MAP = {
    h1: {
        size: "36",
        weight: MEDIUM,
        lineHeight: "44",
    },
    h2: {
        size: "26",
        weight: BOLD,
        lineHeight: "32",
    },
    h3: {
        size: "22",
        weight: BOLD,
        lineHeight: "28",
    },
    h4: {
        size: "20",
        weight: BOLD,
        lineHeight: "24",
    },
    h5: {
        size: "16",
        weight: BOLD,
        lineHeight: "20",
    },
    h6: {
        size: "14",
        weight: BOLD,
        lineHeight: "18",
    },
    subtitle1: {
        size: "20",
        weight: MEDIUM,
        lineHeight: "24",
    },
    subtitle2: {
        size: "16",
        weight: MEDIUM,
        lineHeight: "20",
    },
    subtitle3: {
        size: "14",
        weight: MEDIUM,
        lineHeight: "18",
    },
    body1: {
        size: "20",
        weight: REGULAR,
        lineHeight: "26",
    },
    body2: {
        size: "16",
        weight: REGULAR,
        lineHeight: "22",
    },
    body3: {
        size: "14",
        weight: REGULAR,
        lineHeight: "18",
    },
    body4: {
        size: "12",
        weight: REGULAR,
        lineHeight: "16",
    },
    caption1: {
        size: "16",
        weight: BOLD,
        lineHeight: "20",
        props: {
            textTransform: "uppercase",
        },
    },
    caption2: {
        size: "12",
        weight: BOLD,
        lineHeight: "16",
        props: {
            textTransform: "uppercase",
        },
    },
    overline: {
        size: "10",
        weight: REGULAR,
        lineHeight: "10",
        props: {
            textTransform: "uppercase",
            letterSpacing: "1.5px",
        },
    },
    input: {
        size: "14",
        weight: REGULAR,
        lineHeight: "18",
    },
};
export const VARIANTS = Object.keys(STYLE_MAP);
const StyledTypography = styled("div")(props => {
    var _a;
    return ({
        color: props.$color,
        fontSize: `${STYLE_MAP[props.$variant].size}px`,
        fontWeight: STYLE_MAP[props.$variant].weight,
        lineHeight: `${STYLE_MAP[props.$variant].lineHeight}px`,
        ...(((_a = STYLE_MAP[props.$variant]) === null || _a === void 0 ? void 0 : _a.props) || {}),
        ...(props.noWrap
            ? {
                whiteSpace: "nowrap",
                textOverflow: "ellipsis",
                overflow: "hidden",
            }
            : {}),
    });
});
const Typography = ({ variant, children, color = "#0D1030", ...props }) => (React.createElement(StyledTypography, Object.assign({ "$variant": variant, "$color": color }, props), children));
export { StyledTypography, Typography };
